dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -57,81 +59,113 @@ void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
     }
 }
 
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::prepare(const Block & sample_block)
+{
+    /// Materialize sample_block so that header and reserved scatterColumns are full columns
+    /// Because ser/der don't support constant columns now
+    std::vector<Block> sample_blocks;
+    sample_blocks.push_back(sample_block);
+    HashBaseWriterHelper::materializeBlocks(sample_blocks);
+    header = sample_blocks.back().cloneEmpty();
+    num_columns = header.columns();
+    // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
+    num_bucket = partition_num * fine_grained_shuffle_stream_count;
+    partition_key_containers_for_reuse.resize(collators.size());
+    resetScatterColumns();
+    prepared = true;
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::flush()
+{
+    if (rows_in_blocks > 0)
+        batchWriteFineGrainedShuffle<false>();
+}
+
 template <class StreamWriterPtr>
 void FineGrainedShuffleWriter<StreamWriterPtr>::write(const Block & block)
 {
+    RUNTIME_CHECK_MSG(prepared, "FineGrainedShuffleWriter should be prepared before writing.");
     RUNTIME_CHECK_MSG(
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
+
     size_t rows = block.rows();
     rows_in_blocks += rows;
     if (rows > 0)
     {
         blocks.push_back(block);
     }
 
-    if (static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size)
+    if (blocks.size() == fine_grained_shuffle_stream_count || static_cast<UInt64>(rows_in_blocks) >= batch_send_row_limit)
```

Comment:
1. So we changed the meaning of `fine_grained_shuffle_batch_size`, maybe need to talk to PM and also change the [doc](https://docs.pingcap.com/zh/tidb/dev/system-variables#tiflash_fine_grained_shuffle_batch_size-%E4%BB%8E-v620-%E7%89%88%E6%9C%AC%E5%BC%80%E5%A7%8B%E5%BC%95%E5%85%A5)
2. Just a question: why trigger sending data when blocks.size() equals to fine_grained_shuffle_stream_count

Reply:
Good suggestion and question. For 1, I'll follow. For 2, just think like this:
Previously, no fine_grained_shuffle_stream, N partitions, then one block => N blocks
With fine_grained_shuffle_stream, N partitions, M fine_grained_shuffle_stream_count,  then one block => N * M blocks, thus here use blocks.size() == stream_count, can make final block close to previous block size.

Reply:
Got it. So compared to the original situation, will trigger sending more frequent when blocks are small. Guess this is the reason why `long tail problem` can be fix?

Reply:
Long tail problem is actually fixed by the "flush" interface, which will be invoked concurrently in different threads. Previously, the remain cached blocks that are below thres, will be flushed in "finishWrite" interface, which is invoked sequentially in a single thread.